// seqproto/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDes = exports.createSer = void 0;
function createSer() {
    var buffer = new ArrayBuffer(Math.pow(2, 24));
    return {
        index: 0,
        buffer: buffer,
        uint8Array: new Uint8Array(buffer),
        uint32Array: new Uint32Array(buffer),
        float32Array: new Float32Array(buffer),
        serializeBoolean: serializeBoolean,
        serializeUInt32: serializeUInt32,
        serializeFloat32: serializeFloat32,
        serializeString: serializeString,
        serializeArray: serializeArray,
        getBuffer: function () { return this.buffer.slice(0, this.index * 4); }
    };
}
exports.createSer = createSer;
function createDes(buffer) {
    var n32 = Math.floor(buffer.byteLength / 4);
    return {
        index: 0,
        buffer: buffer,
        uint8Array: new Uint8Array(buffer),
        uint32Array: new Uint32Array(buffer, 0, n32),
        float32Array: new Float32Array(buffer, 0, n32),
        setBuffer: function (buffer) {
            var n32 = Math.floor(buffer.byteLength / 4);
            this.buffer = buffer;
            this.index = 0;
            this.uint8Array = new Uint8Array(buffer);
            this.uint32Array = new Uint32Array(buffer, 0, n32);
            this.float32Array = new Float32Array(buffer, 0, n32);
        },
        deserializeBoolean: deserializeBoolean,
        deserializeUInt32: deserializeUInt32,
        deserializeFloat32: deserializeFloat32,
        deserializeString: deserializeString,
        deserializeArray: deserializeArray
    };
}
exports.createDes = createDes;
function serializeBoolean(b) {
    this.uint32Array[this.index++] = b ? 1 : 0;
}
function deserializeBoolean() {
    return this.uint32Array[this.index++] === 1;
}
function serializeUInt32(n) {
    this.uint32Array[this.index++] = n;
}
function deserializeUInt32() {
    return this.uint32Array[this.index++];
}
function serializeFloat32(n) {
    this.float32Array[this.index++] = n;
}
function deserializeFloat32() {
    return this.float32Array[this.index++];
}
var textEncoder = new TextEncoder();
function serializeString(str) {
    var r = textEncoder.encodeInto(str, new Uint8Array(this.buffer, (this.index + 1) * 4));
    this.uint32Array[this.index] = r.written;
    this.index += Math.ceil(r.written / 4) + 1;
}
var textDecoder = new TextDecoder();
function deserializeString() {
    var len = this.uint32Array[this.index++];
    var decoded = textDecoder.decode(new Uint8Array(this.buffer, this.index * 4, len));
    this.index += Math.ceil(len / 4);
    return decoded;
}
function serializeArray(arr, serialize) {
    var len = arr.length;
    this.serializeUInt32(len);
    for (var i = 0; i < len; i++) {
        serialize(this, arr[i]);
    }
}
function deserializeArray(deserialize) {
    var len = this.deserializeUInt32();
    var arr = new Array(len);
    for (var i = 0; i < len; i++) {
        arr[i] = deserialize(this);
    }
    return arr;
}
