// @orama/plugin-vitepress/dist/index.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    }, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(g && (g = 0, op[0] && (_ = 0)), _)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import MarkdownIt from 'markdown-it';
import { JSDOM } from 'jsdom';
import { presets } from '@orama/searchbox';
import { create, insertMultiple } from '@orama/orama';
import { persist } from '@orama/plugin-data-persistence';
import slugify from 'slugify';
import { readFileSync } from 'fs';
var md = new MarkdownIt({
    html: true
});
function createOramaContentLoader(paths, root, base) {
    return _async_to_generator(function() {
        var contents, db;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    contents = paths.map(function(file) {
                        return {
                            path: file.replace(root, '').replace('.md', ''),
                            html: md.render(readFileSync(file, 'utf-8'), '')
                        };
                    }).map(parseHTMLContent).flatMap(function(data) {
                        return formatForOrama(data, base);
                    });
                    return [
                        4,
                        create({
                            schema: presets.docs.schema
                        })
                    ];
                case 1:
                    db = _state.sent();
                    return [
                        4,
                        insertMultiple(db, contents)
                    ];
                case 2:
                    _state.sent();
                    return [
                        2,
                        persist(db, 'json', 'browser')
                    ];
            }
        });
    })();
}
function parseHTMLContent(param) {
    var html = param.html, path = param.path;
    var dom = new JSDOM(html);
    var document = dom.window.document;
    var sections = [];
    var headers = document.querySelectorAll('h1, h2, h3, h4, h5, h6');
    headers.forEach(function(header) {
        var sectionTitle = header.textContent.trim();
        var headerTag = header.tagName.toLowerCase();
        var sectionContent = '';
        var sibling = header.nextElementSibling;
        while(sibling && ![
            'H1',
            'H2',
            'H3',
            'H4',
            'H5',
            'H6'
        ].includes(sibling.tagName)){
            sectionContent += sibling.textContent.trim() + '\n';
            sibling = sibling.nextElementSibling;
        }
        sections.push({
            title: sectionTitle,
            header: headerTag,
            content: sectionContent,
            path: path
        });
    });
    return sections;
}
function formatForOrama(data, base) {
    try {
        var firstH1Header = data.find(function(section) {
            return section.header === 'h1';
        });
        return data.map(function(res) {
            return {
                title: res.title,
                content: res.content,
                section: ((firstH1Header === null || firstH1Header === void 0 ? void 0 : firstH1Header.title) || '').replace(/\s$/, ''),
                path: base + (res === null || res === void 0 ? void 0 : res.path) + '#' + slugify.default(res.title, {
                    lower: true
                }),
                category: ''
            };
        });
    } catch (error) {
        console.error(error);
        return [];
    }
}
function removeTrailingSlash(value) {
    return value.endsWith('/') ? value.slice(0, -1) : value;
}
export function OramaPlugin() {
    var pluginOptions = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {};
    var resolveConfig;
    var virtualModuleId = 'virtual:search-data';
    var resolvedVirtualModuleId = "\0".concat(virtualModuleId);
    return {
        name: 'oramasearch',
        enforce: 'pre',
        configResolved: function configResolved(config) {
            if (resolveConfig) {
                return;
            }
            resolveConfig = config;
            var vitepressConfig = config.vitepress;
            if (!vitepressConfig) {
                return;
            }
            var selfBuildEnd = vitepressConfig.buildEnd;
            vitepressConfig.buildEnd = function(siteConfig) {
                var _pluginSiteConfig_buildEnd;
                selfBuildEnd === null || selfBuildEnd === void 0 ? void 0 : selfBuildEnd(siteConfig);
                siteConfig = Object.assign(siteConfig || {});
                pluginSiteConfig === null || pluginSiteConfig === void 0 ? void 0 : (_pluginSiteConfig_buildEnd = pluginSiteConfig.buildEnd) === null || _pluginSiteConfig_buildEnd === void 0 ? void 0 : _pluginSiteConfig_buildEnd.call(pluginSiteConfig, siteConfig);
            };
            var selfTransformHead = vitepressConfig.transformHead;
            vitepressConfig.transformHead = function(ctx) {
                return _async_to_generator(function() {
                    var _pluginSiteConfig_transformHead, selfHead, pluginHead;
                    return _ts_generator(this, function(_state) {
                        switch(_state.label){
                            case 0:
                                return [
                                    4,
                                    Promise.resolve(selfTransformHead === null || selfTransformHead === void 0 ? void 0 : selfTransformHead(ctx))
                                ];
                            case 1:
                                selfHead = _state.sent() || [];
                                return [
                                    4,
                                    Promise.resolve(pluginSiteConfig === null || pluginSiteConfig === void 0 ? void 0 : (_pluginSiteConfig_transformHead = pluginSiteConfig.transformHead) === null || _pluginSiteConfig_transformHead === void 0 ? void 0 : _pluginSiteConfig_transformHead.call(pluginSiteConfig, ctx))
                                ];
                            case 2:
                                pluginHead = _state.sent() || [];
                                return [
                                    2,
                                    selfHead.concat(pluginHead)
                                ];
                        }
                    });
                })();
            };
        },
        config: function() {
            return {
                resolve: {
                    alias: {
                        './VPNavBarSearch.vue': new URL('./Search.vue', import.meta.url).pathname
                    }
                }
            };
        },
        resolveId: function resolveId(id) {
            return _async_to_generator(function() {
                return _ts_generator(this, function(_state) {
                    if (id === virtualModuleId) {
                        return [
                            2,
                            resolvedVirtualModuleId
                        ];
                    }
                    return [
                        2
                    ];
                });
            })();
        },
        load: function load(id) {
            return _async_to_generator(function() {
                var _resolveConfig_vitepress_userConfig, root, _resolveConfig_vitepress_userConfig_base, base, pages, _, _1, _2;
                return _ts_generator(this, function(_state) {
                    switch(_state.label){
                        case 0:
                            if (id !== resolvedVirtualModuleId) return [
                                2
                            ];
                            root = resolveConfig.vitepress.srcDir || resolveConfig.vitepress.root;
                            base = removeTrailingSlash((_resolveConfig_vitepress_userConfig_base = (_resolveConfig_vitepress_userConfig = resolveConfig.vitepress.userConfig) === null || _resolveConfig_vitepress_userConfig === void 0 ? void 0 : _resolveConfig_vitepress_userConfig.base) !== null && _resolveConfig_vitepress_userConfig_base !== void 0 ? _resolveConfig_vitepress_userConfig_base : '');
                            pages = resolveConfig.vitepress.pages.map(function(page) {
                                return "".concat(root, "/").concat(page);
                            });
                            _1 = (_ = "\n        const data = ").concat;
                            _2 = JSON.stringify;
                            return [
                                4,
                                createOramaContentLoader(pages, root, base)
                            ];
                        case 1:
                            return [
                                2,
                                _1.apply(_, [
                                    _2.apply(JSON, [
                                        _state.sent()
                                    ]),
                                    ";\n        const analytics = "
                                ]).concat(JSON.stringify(pluginOptions.analytics), ";\n        export default { data, analytics };\n      ")
                            ];
                    }
                });
            })();
        }
    };
}
var pluginSiteConfig = {
    buildEnd: function buildEnd() {},
    transformHead: function transformHead() {
        return [];
    }
};
